plutus-core/testlib/PlutusCore/Generators/PIR.hs

```diff
@@ -0,0 +1,1752 @@
+                         [(k, TyFun () a b) | (_, b) <- shrinkKindAndType ctx (k, b)]
+    -- This case needs to be handled with a bit of care. First we shrink applications by
+    -- doing simple stuff like shrinking the function and body separately when we can.
+    -- The slightly tricky case is the concat trace. See comment below.
+    TyApp _ f a       -> [(ka, a) | ka `leKind` k] ++
+                         [(k, b)                     | TyLam _ x _ b <- [f], not $ Set.member x (fvType b)] ++
+                         [(k, substClosedType x a b) | TyLam _ x _ b <- [f], null (fvType a)] ++
+                         -- Here we try to shrink the function f, if we get something whose kind
+                         -- is small enough we can return the new function f', otherwise we
+                         -- apply f' to `fixKind ctx a ka'` - which takes `a` and tries to rewrite it
+                         -- to something of kind `ka'`.
+                         concat [case kf' of
+                                   Type{}              -> [(kf', f')]
+                                   KindArrow _ ka' kb' -> [ (kb', TyApp () f' (fixKind ctx a ka'))
+                                                          | leKind kb' k, leKind ka' ka]
+                                 | (kf', f') <- shrinkKindAndType ctx (KindArrow () ka k, f)] ++
+                         -- Here we shrink the argument and fixup the function to have the right kind.
+                         [(k, TyApp () (fixKind ctx f (KindArrow () ka' k)) a)
+                         | (ka', a) <- shrinkKindAndType ctx (ka, a)]
+      where ka = inferKind_ ctx a
+    -- type lambdas shrink by either shrinking the kind of the argument or shrinking the body
+    TyLam _ x ka b    -> [ (KindArrow () ka' kb, TyLam () x ka' $ substClosedType x (minimalType ka) b)
+                         | ka' <- shrink ka] ++
+                         [ (KindArrow () ka kb', TyLam () x ka b)
+                         | (kb', b) <- shrinkKindAndType (Map.insert x ka ctx) (kb, b)]
+      where KindArrow _ _ kb = k
+    TyForall _ x ka b -> [ (k, b) | not $ Set.member x (fvType b) ] ++
+                         -- (above) If the bound variable doesn't matter we get rid of the binding
+                         [ (k, TyForall () x ka' $ substClosedType x (minimalType ka) b)
+                         | ka' <- shrink ka] ++
+                         -- (above) we can always just shrink the bound variable to a smaller kind
+                         -- and ignore it
+                         [ (k, TyForall () x ka b)
+                         | (_, b) <- shrinkKindAndType (Map.insert x ka ctx) (Star, b)]
+                         -- (above) or we shrink the body
+    TyBuiltin{}       -> []
+    TyIFix{}          -> error "shrinkKindAndType: TyIFix"
+
+-- CODE REVIEW: does this exist anywhere?
+inferKind :: Map TyName (Kind ()) -> Type TyName DefaultUni () -> Maybe (Kind ())
```

Comment:
https://github.com/input-output-hk/plutus/blob/master/plutus-core/plutus-core/src/PlutusCore/TypeCheck/Internal.hs#L212

I feel like "in the typechecker" is not a very hidden location ;)

Reply:
Sure... Just have to expose that module in the cabal file...

Reply:
> Sure... Just have to expose that module in the cabal file...

I did it for you a few days ago. Don't use right now, though, as I've simplified the interface a little, but the PR is not merged yet. I'll update your PR once mine is merged.

Also do feel free to expose modules in the cabal file. Maybe we should go ahead and just expose all of them.